sui/contract/sources/constant_product.move

```diff
@@ -7,78 +7,133 @@
 /// As well as Sui's reference implementation:
 ///     https://github.com/MystenLabs/sui/blob/devnet/sui_programmability/examples/defi/sources/pool.move
 /// 
-/// TODO: how to enforce single pool for a pair without permissioning?
 /// TODO: add more tests for different decimals
 /// TODO: add tests for edge cases
 /// TODO: slippage when adding liquidity
 /// TODO: burn initial liquidity
 module aux::constant_product {
-    use sui::object::{Self, UID};
-    use sui::coin::{Self, Coin};
+    use std::type_name::{Self, TypeName};
+
     use sui::balance::{Self, Supply};
-    use sui::transfer;
+    use sui::coin::{Self, Coin};
+    use sui::event;
     use sui::math;
+    use sui::object::{Self, ID, UID};
+    use sui::transfer;
     use sui::tx_context::{Self, TxContext};
+    use sui::vec_set::{Self, VecSet};
+
+    /*************/
+    /* Constants */
+    /*************/
 
-    const EInvalidArg: u64 = 0;
-    const EPoolEmpty: u64 = 1;
-    const EPoolOverflow: u64 = 2;
-    const EMinSwap: u64 = 3;
+    const EBadWitness: u64 = 0;
+    const EInvalidArg: u64 = 1;
+    const EPoolAlreadyExists: u64 = 2;
+    const EPoolEmpty: u64 = 3;
+    const EPoolOverflow: u64 = 4;
+    const EMinSwap: u64 = 5;
 
     const MAX_BPS: u128 = 10000;
 
-    struct LP<phantom X, phantom Y, phantom W> has drop {}
+    /***********/
+    /* Objects */
+    /***********/
+
+    struct Pools has key {
+        id: UID,
+        pools: VecSet<TypeName>
+    }
 
-    struct Pool<phantom X, phantom Y, phantom W> has key {
+    struct Pool<phantom X, phantom Y> has key {
         id: UID,
         reserve_x: Coin<X>,
         reserve_y: Coin<Y>,
-        supply_lp: Supply<LP<X, Y, W>>,
-        fee_bps: u64,
+        supply_lp: Supply<LP<X, Y>>,
+        fee_bps: u64
+    }
+
+    struct LP<phantom X, phantom Y> has drop {}
+
+    /**********/
+    /* Events */
+    /**********/
+
+    struct PoolsCreated has copy, drop {
+        id: ID
     }
 
-    public fun reserve_x<X, Y, W>(self: &Pool<X, Y, W>): &Coin<X> {
+    struct PoolCreated<phantom X, phantom Y> has copy, drop {
+        id: ID
+    }
+
+    struct Swapped<phantom IN, phantom OUT> has copy, drop {
```

Comment:
Is it conventional to use all caps for type args?

Reply:
i'm just basing off Sui's official examples. they use single letter `T` for type parameters, `ALL_CAPS` for type arguments. this should probably actually be `<I, O>`